pkg/gce-pd-csi-driver/node_test.go

```diff
@@ -378,3 +387,75 @@ func TestNodeGetCapabilities(t *testing.T) {
 		t.Fatalf("Unexpedted error: %v", err)
 	}
 }
+
+func TestConcurrentNodeOperations(t *testing.T) {
+	okToTest := make(chan struct{}, 2)
+	opStarted := make(chan struct{}, 2)
+	reqFinished := make(chan struct{}, 2)
+
+	gceDriver := getTestBlockingGCEDriver(t, okToTest, opStarted)
+	ns := gceDriver.ns
+	vol1PublishTargetAReq := &csi.NodePublishVolumeRequest{
+		VolumeId:          defaultVolumeID + "1",
+		TargetPath:        defaultTargetPath + "a",
+		StagingTargetPath: defaultStagingPath + "1",
+		Readonly:          false,
+		VolumeCapability:  stdVolCap,
+	}
+	vol1PublishTargetBReq := &csi.NodePublishVolumeRequest{
+		VolumeId:          defaultVolumeID + "1",
+		TargetPath:        defaultTargetPath + "b",
+		StagingTargetPath: defaultStagingPath + "1",
+		Readonly:          false,
+		VolumeCapability:  stdVolCap,
+	}
+	vol2PublishTargetCReq := &csi.NodePublishVolumeRequest{
+		VolumeId:          defaultVolumeID + "2",
+		TargetPath:        defaultTargetPath + "c",
+		StagingTargetPath: defaultStagingPath + "2",
+		Readonly:          false,
+		VolumeCapability:  stdVolCap,
+	}
+
+	runRequestInBackground := func(req *csi.NodePublishVolumeRequest) {
+		_, err := ns.NodePublishVolume(context.Background(), req)
+		if err != nil {
+			t.Fatalf("Unexpedted error: %v", err)
+		}
+		reqFinished <- struct{}{}
+	}
+
+	// Start first valid request vol1PublishTargetAReq
+	go runRequestInBackground(vol1PublishTargetAReq)
+
+	// Blocks until vol1PublishTargetAReq runs until the Mount
+	<-opStarted
+
+	// Check that vol1PublishTargetBReq is rejected, due to same volume ID
+	_, err := ns.NodePublishVolume(context.Background(), vol1PublishTargetBReq)
+	if err != nil {
+		serverError, ok := status.FromError(err)
+		if !ok {
+			t.Fatalf("Could not get error status code from err: %v", err)
+		}
+		if serverError.Code() != codes.Aborted {
+			t.Fatalf("Expected error code: %v, got: %v. err : %v", codes.Aborted, serverError.Code(), err)
+		}
+	} else {
+		t.Fatalf("Expected error: %v, got no error", codes.Aborted)
+	}
+
+	// Start second valid request vol2PublishTargetCReq
+	go runRequestInBackground(vol2PublishTargetCReq)
+
+	// Check that vol2PublishTargetCReq is able to run
+	<-opStarted
```

Comment:
in the case that `vol2PublishTargetCReq` is actually unable to run and fails for whatever reason won't this block forever?

Reply:
I changed it so that failures should not block anymore.